### ISSUE TITLE
Offline checking is oversensitive / needs a 'try again now' option

### ISSUE THREAD

**Issue by rjmackay** [Reviewable/Reviewable#321]
The offline status when reviewing has far too many false positives for me. At best guess my wifi is dropping occasionally so the Offline modal pops up. This is far too sensitive given everything else continues to work (ie. slack is connected, other apps working fine etc)
However this is made worse because there's no way to trigger a reconnect without turning wifi on/off again. Can you add a "retry now" button? 
Ideally make the offline check less sensitive too.

**Comment by pkaminski** [comment]
I've been tracking this problem for a while and at this point I'm pretty sure there's something fishy going on, but I don't think it's as simple as needing to force retries.  I instrumented the overlay to report some more info to me, and added a "reason" line to the offline overlay -- could you let me know what the reason is next time you end up offline?  I'd like to get to the bottom of this.  Thanks.

**Comment by rjmackay** [comment]
Cool. I can do that. I'll pass on the request to my co-workers too.

A few things I noticed:
- the web socket connection appeared to be failing pretty much continuously
- Everything resolved when I reloaded the page.
- All the comments/etc I had written were saved locally when I reloaded the page. So from a user perspective just knowing that I could reload and then hit 'Publish' would have been great. But the messaging suggested that I would lose data.

**Comment by pkaminski** [comment]
Huh, if the websocket is failing all the time, could you turn on `Firebase.enableLogging(true)` in the console and send a representative chunk of that to me?  Unless you're behind some draconian firewall or proxy, it really shouldn't be happening and will have pretty bad effects on performance.

The messaging suggests that you might lose data because all Reviewable knows is that _some_ write _may_ have failed (like an unconfirmed transaction -- there's no way to tell if it went through or not), but the all-purpose catcher of this condition isn't smart enough to figure out how much it might matter.  This could be improved, of course, but the idea was that this should happen extremely rarely in the first place...  In general, Reviewable saves data very often (either immediately or after a 100ms delay), so unless a lot of writes are pending (new new overlay tells you the number) you should be fine to reload without much or any loss.

**Comment by kballard** [comment]
This just happened to me again (though it's the first time today, so that's an improvement). I didn't have logging turned on, but the offline dialog says "Reason: 1 stalled write". I noticed the connection indicator appear while I was writing a comment, so I assume the stalled write was in trying to save a comment draft, but my comment appears intact after reloading. Could Reviewable have saved a later draft, thereby making the write for the earlier draft obsolete, and therefore going offline because that old write stalled out is unnecessary?

**Comment by pkaminski** [comment]
Arrgh, the mystery is compounding.  I wanted to look up your event in my tracking system, but I only see one at 10:05am PDT, whereas your comment's timestamp above indicates it happened closer to 3pm PDT -- is that right?  Now I'm left wondering if my capture system is also broken somehow, or dropping events...

The rest of your analysis is plausible.  I don't check for overwrites because 1) it's a bit tricky due to Firebase's API and 2) *every* write should eventually invoke the callback, no matter what.  It's also possible that the stalled write was a "background write", like the permission ticket refresh request that goes out every 15 minutes, etc.

**Comment by kballard** [comment]
Yeah, I filed the comment right after the issue happened, so that would have been just after 3PM PDT.

**Comment by kballard** [comment]
Just had it happen again. The red indicator showed up while I was writing a comment. I then hit Publish, and the Publish button showed the activity indicator for something like 8 seconds, then it appeared to finish publishing, even though the red indicator was still visible. A few seconds later I got the Offline overlay with the reason "1 pending write (1 stalled)".

This happened around 3:58 PM PDT.

After reloading, my comment had indeed been published properly.

**Comment by SpainTrain** [comment]
I am seeing this issue fairly regularly.  I can largely rule out firewall issues because it happens at home and work.  The `Reason` is the same as mentioned elsewhere in the thread `Reason: 2 pending writes (2 stalled)`.  Similarly, whenever this happens, I refresh and have not lost any data.  I will begin to run with firebase debugging enabled and provide info from that.

I just experienced it at ~16:20 EDT.

UA: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.103 Safari/537.36

**Comment by pkaminski** [comment]
I deployed a small tweak to the code to test a low-likelihood hypothesis, but I'm still basically stumped.  More data would be very welcome.

**Comment by SpainTrain** [comment]
Somewhat amusingly, it appears that enabling firebase debugging affects the behavior of the bug.  The issue has not happened in a few days since I have been running with firebase debugging enabled.  Even more interestingly, I just hit the bug for a review in which I forgot to enable firebase debugging and turning debugging on actually made the Offline overlay go away.

Unfortunately this means that debugging to find root cause may be even harder.  Are there any other things we could capture that would be helpful in nailing down the cause?

**Comment by pkaminski** [comment]
Uh?...  Are you literally saying that:
1. The offline overlay popped up, complaining about stalled writes.
2. You ran `Firebase.enableLogging(true)` in the console.
3. The overlay immediately disappeared, without reloading the page.

If so, that would be a huge clue in an of itself, if I could only figure out what it means.

**Comment by SpainTrain** [comment]
> Are you literally saying that

yes, your description is accurate

> if I could only figure out what it means

:smile: might firebase support be able to provide some insights?

**Comment by pkaminski** [comment]
Let's make sure this wasn't a freak coincidence first.  Could you (or anyone else) try running `Firebase.enableLogging(true)` from the console when you next get the offline overlay and confirm that this makes the overlay disappear?

Firebase support isn't likely to be much help thanks to their recent policy of "give us a clean repro or go away" (@startupandrew, in case he's paying attention).  Luckily, I'm still on a version of the Firebase SDK that provides full source code, though, so I can go spelunking myself.  A first pass through `enableLogging` doesn't show how that function could unblock anything, however, as it just sets some flags...

**Comment by FichteFoll** [comment]
For me, this bug surfaces in the following form:

1. I write text.
2. My connection drops (which does happen occasionally, no WiFi)
3. After a while reviewable opens the modal complaining about having no connection. The exact reason is (this time): "Reason: no connection to Firebase, 73 pending writes (73 stalled)"
4. After a short while (usually less than 30s) I regain connection and all my other applications reconnect (Skype, Pidgin, IRC (if that even dropped at all)), but reviewable **does not**! The modal stays and stays.

After enabling logging I get a lot of this:

```
p:0: Window became visible.  Reducing delay.   9c254ca0.libs.js:13:11086
p:0: Making a connection attempt   9c254ca0.libs.js:13:11086
c:0:33: Connection created   9c254ca0.libs.js:13:11086
"c:0:33:0 Connecting via long-poll to https://reviewable.firebaseio.com/.lp?start=t&ser=79919477&cb=45&v=5&ls=jEXOuzWy2nmrasM6Mjr47EytG9TjVjdB  " 9c254ca0.libs.js:13:11086
"Long-poll script failed to load: https://reviewable.firebaseio.com/.lp?start=t&ser=79919477&cb=45&v=5&ls=jEXOuzWy2nmrasM6Mjr47EytG9TjVjdB " 9c254ca0.libs.js:13:11086
c:0:33:0 Longpoll is closing itself   9c254ca0.libs.js:13:11086
c:0:33: Realtime connection failed.   9c254ca0.libs.js:13:11086
c:0:33: Closing realtime connection.   9c254ca0.libs.js:13:11086
c:0:33: Shutting down all connections   9c254ca0.libs.js:13:11086
p:0: data client disconnected   9c254ca0.libs.js:13:11086
p:0: Trying to reconnect in 0ms   9c254ca0.libs.js:13:11086
0: onDisconnectEvents   9c254ca0.libs.js:13:11086
p:0: Making a connection attempt   9c254ca0.libs.js:13:11086
c:0:34: Connection created   9c254ca0.libs.js:13:11086
"c:0:34:0 Connecting via long-poll to https://reviewable.firebaseio.com/.lp?start=t&ser=43129628&cb=46&v=5&ls=jEXOuzWy2nmrasM6Mjr47EytG9TjVjdB  " 9c254ca0.libs.js:13:11086
"Long-poll script failed to load: https://reviewable.firebaseio.com/.lp?start=t&ser=43129628&cb=46&v=5&ls=jEXOuzWy2nmrasM6Mjr47EytG9TjVjdB " 9c254ca0.libs.js:13:11086
c:0:34:0 Longpoll is closing itself   9c254ca0.libs.js:13:11086
c:0:34: Realtime connection failed.   9c254ca0.libs.js:13:11086
c:0:34: Closing realtime connection.   9c254ca0.libs.js:13:11086
c:0:34: Shutting down all connections   9c254ca0.libs.js:13:11086
p:0: data client disconnected   9c254ca0.libs.js:13:11086
p:0: Trying to reconnect in 261.4411611182722ms   9c254ca0.libs.js:13:11086
0: onDisconnectEvents   9c254ca0.libs.js:13:11086
p:0: Making a connection attempt   9c254ca0.libs.js:13:11086
c:0:35: Connection created   9c254ca0.libs.js:13:11086
"c:0:35:0 Connecting via long-poll to https://reviewable.firebaseio.com/.lp?start=t&ser=73531023&cb=47&v=5&ls=jEXOuzWy2nmrasM6Mjr47EytG9TjVjdB  " 9c254ca0.libs.js:13:11086
"Long-poll script failed to load: https://reviewable.firebaseio.com/.lp?start=t&ser=73531023&cb=47&v=5&ls=jEXOuzWy2nmrasM6Mjr47EytG9TjVjdB " 9c254ca0.libs.js:13:11086
c:0:35:0 Longpoll is closing itself   9c254ca0.libs.js:13:11086
c:0:35: Realtime connection failed.   9c254ca0.libs.js:13:11086
c:0:35: Closing realtime connection.   9c254ca0.libs.js:13:11086
c:0:35: Shutting down all connections   9c254ca0.libs.js:13:11086
p:0: data client disconnected   9c254ca0.libs.js:13:11086
p:0: Trying to reconnect in 56.4832285704071ms   9c254ca0.libs.js:13:11086
0: onDisconnectEvents  
```

Eventually I decide to reload the page so that I can finally continue writing. Only to find out that reviewable now shows the modal before I can even see anything (but after the "checking permission" spinner):

```
p:0: Making a connection attempt   9c254ca0.libs.js:13:11086
c:0:8: Connection created   9c254ca0.libs.js:13:11086
"c:0:8:0 Connecting via long-poll to https://reviewable.firebaseio.com/.lp?start=t&ser=12776196&cb=10&v=5  " 9c254ca0.libs.js:13:11086
"Long-poll script failed to load: https://reviewable.firebaseio.com/.lp?start=t&ser=12776196&cb=10&v=5 " 9c254ca0.libs.js:13:11086
c:0:8:0 Longpoll is closing itself   9c254ca0.libs.js:13:11086
c:0:8: Realtime connection failed.   9c254ca0.libs.js:13:11086
c:0:8: Closing realtime connection.   9c254ca0.libs.js:13:11086
c:0:8: Shutting down all connections   9c254ca0.libs.js:13:11086
p:0: data client disconnected   9c254ca0.libs.js:13:11086
p:0: Window isn't visible.  Delaying reconnect.   9c254ca0.libs.js:13:11086
p:0: Trying to reconnect in 61488.102674422706ms   9c254ca0.libs.js:13:11086
0: onDisconnectEvents   9c254ca0.libs.js:13:11086
p:0: Window became visible.  Reducing delay.   9c254ca0.libs.js:13:11086
p:0: Making a connection attempt   9c254ca0.libs.js:13:11086
c:0:9: Connection created   9c254ca0.libs.js:13:11086
"c:0:9:0 Connecting via long-poll to https://reviewable.firebaseio.com/.lp?start=t&ser=79192989&cb=11&v=5  " 9c254ca0.libs.js:13:11086
"Long-poll script failed to load: https://reviewable.firebaseio.com/.lp?start=t&ser=79192989&cb=11&v=5 " 9c254ca0.libs.js:13:11086
c:0:9:0 Longpoll is closing itself   9c254ca0.libs.js:13:11086
c:0:9: Realtime connection failed.   9c254ca0.libs.js:13:11086
c:0:9: Closing realtime connection.   9c254ca0.libs.js:13:11086
c:0:9: Shutting down all connections   9c254ca0.libs.js:13:11086
p:0: data client disconnected   9c254ca0.libs.js:13:11086
p:0: Trying to reconnect in 0ms   9c254ca0.libs.js:13:11086
0: onDisconnectEvents   9c254ca0.libs.js:13:11086
p:0: Making a connection attempt   9c254ca0.libs.js:13:11086
c:0:10: Connection created   9c254ca0.libs.js:13:11086
"c:0:10:0 Connecting via long-poll to https://reviewable.firebaseio.com/.lp?start=t&ser=56780094&cb=12&v=5  " 9c254ca0.libs.js:13:11086
"Long-poll script failed to load: https://reviewable.firebaseio.com/.lp?start=t&ser=56780094&cb=12&v=5 " 9c254ca0.libs.js:13:11086
c:0:10:0 Longpoll is closing itself   9c254ca0.libs.js:13:11086
c:0:10: Realtime connection failed.   9c254ca0.libs.js:13:11086
c:0:10: Closing realtime connection.   9c254ca0.libs.js:13:11086
c:0:10: Shutting down all connections   9c254ca0.libs.js:13:11086
p:0: data client disconnected   9c254ca0.libs.js:13:11086
p:0: Trying to reconnect in 298.16206229453434ms   9c254ca0.libs.js:13:11086
0: onDisconnectEvents   9c254ca0.libs.js:13:11086
p:0: Making a connection attempt   9c254ca0.libs.js:13:11086
c:0:11: Connection created   9c254ca0.libs.js:13:11086
"c:0:11:0 Connecting via long-poll to https://reviewable.firebaseio.com/.lp?start=t&ser=81360989&cb=13&v=5  " 9c254ca0.libs.js:13:11086
"Long-poll script failed to load: https://reviewable.firebaseio.com/.lp?start=t&ser=81360989&cb=13&v=5 " 9c254ca0.libs.js:13:11086
c:0:11:0 Longpoll is closing itself   9c254ca0.libs.js:13:11086
c:0:11: Realtime connection failed.   9c254ca0.libs.js:13:11086
c:0:11: Closing realtime connection.   9c254ca0.libs.js:13:11086
c:0:11: Shutting down all connections   9c254ca0.libs.js:13:11086
p:0: data client disconnected   9c254ca0.libs.js:13:11086
p:0: Trying to reconnect in 549.5014647728154ms   9c254ca0.libs.js:13:11086
0: onDisconnectEvents   9c254ca0.libs.js:13:11086
p:0: Making a connection attempt   9c254ca0.libs.js:13:11086
c:0:12: Connection created   9c254ca0.libs.js:13:11086
"c:0:12:0 Connecting via long-poll to https://reviewable.firebaseio.com/.lp?start=t&ser=27427886&cb=14&v=5  " 9c254ca0.libs.js:13:11086
"Long-poll script failed to load: https://reviewable.firebaseio.com/.lp?start=t&ser=27427886&cb=14&v=5 " 9c254ca0.libs.js:13:11086
c:0:12:0 Longpoll is closing itself   9c254ca0.libs.js:13:11086
c:0:12: Realtime connection failed.   9c254ca0.libs.js:13:11086
c:0:12: Closing realtime connection.   9c254ca0.libs.js:13:11086
c:0:12: Shutting down all connections   9c254ca0.libs.js:13:11086
p:0: data client disconnected   9c254ca0.libs.js:13:11086
p:0: Trying to reconnect in 1357.790163980703ms   9c254ca0.libs.js:13:11086
0: onDisconnectEvents  
```

**Comment by pkaminski** [comment]
This looks like a separate issue -- your machine can't connect to the Firebase server at all, for some reason.  Can you independently check your connectivity to s-usc1c-nss-130.firebaseio.com?

**Comment by FichteFoll** [comment]
@pkaminski I could connect, but it didn't display anything. So I allowed `firebase.io` in noscript and now connecting in reviewable works as well. Thanks for the pointer, although it wasn't what you expected.

I wonder why it worked previously.

**Comment by pkaminski** [comment]
Glad to hear you're back in business.  I guess something must've changed, either on the Firebase side or in Noscript, but I have no idea what.

**Comment by SpainTrain** [comment]
I cannot reproduce the fix regarding enabling firebase debugging, so perhaps that was just a coincidence.

I just encountered this again and have a warning in the console

![selection_137](https://cloud.githubusercontent.com/assets/380032/16740293/653cdbaa-476c-11e6-97ff-fa21cc1a3674.png)

**Comment by pkaminski** [comment]
I'm despairing of finding the root cause for the writes that never complete so I've changed the algorithm to discount any stalled writes that are followed by successful one.  These stalled writes will still cause a red warning indicator to appear in the lower right corner but will no longer lock out the UI with an offline overlay.

I'll be tracking changes in the event's metrics but it'll take a few days before use of the new release is widespread.  In the meantime, I'd be interested in your personal impressions of whether this helped or not.  Thanks!

**Comment by sherm1** [comment]
Per chat discussion, we have been seeing "Offline" messages frequently today while attempting to write review comments on modest-sized PRs. I'm getting them in Palo Alto, and @david-german-tri is getting them in Cambridge, MA. I'm on a hardwired high speed line.

**Comment by pkaminski** [comment]
The previous algorithm change was ineffective, but I finally got a repro on my own machine, in full dev configuration no less.  It really does look like a bug in the Firebase SDK where the connection essentially dies but the socket remains open and is treated as "connected".  I added a potential workaround that tries to detect the condition, do a test write, and eventually bounces the connection to get things moving again.  If it works (and your network latency is low enough) there should be no disruption whatsoever.  Fingers crossed...

**Comment by pkaminski** [comment]
With the latest workaround, stale write offlines have been reduced by 2-3x or so.  Perhaps at this point they're actually mostly true positives?

**Comment by kballard** [comment]
Lately, when I switch to my pinned tab for my reviews page I usually see the Offline banner but it vanishes after a second, not really sure what's going on there. When actually writing reviews the Offline banners seem to have gone away entirely. So this is great!

**Comment by SpainTrain** [comment]
> Perhaps at this point they're actually mostly true positives?

FWIW, I used to encounter this issue routinely and have not seen it for at least a couple weeks.  Anecdotal, but possibly the latest workaround is to thank (so thanks!).

**Comment by pkaminski** [comment]
Given the stats and feedback, and in a wild fit of optimism, I'm going to close this issue.  😄
